src/databricks/labs/ucx/assessment/jobs.py

```diff
@@ -104,3 +116,221 @@ def snapshot(self) -> Iterable[JobInfo]:
+    """
+    if task.new_cluster is not None:
+        return task.new_cluster
+    return None
+
+
+@dataclass
+class SubmitRunInfo:
+    run_ids: str  # JSON-encoded list of run ids
+    hashed_id: str  # a pseudo id that combines all the hashable attributes of the run
+    failures: str = "[]"  # JSON-encoded list of failures
+
+
+_FS_LEVEL_KEY_REGEXP = [
+    "*.fs.s3a.*",
+    "*.fs.azure.*",
+    "*.fs.wasb.*",
+    "*.fs.abfs.*",
+    "*.fs.adl.*",
+]
+
+
+class SubmitRunsCrawler(CrawlerBase[SubmitRunInfo], JobsMixin, CheckClusterMixin):
+
+    def __init__(self, ws: WorkspaceClient, sbe: SqlBackend, schema):
+        super().__init__(sbe, "hive_metastore", schema, "submit_runs", SubmitRunInfo)
+        self._ws = ws
+
+    def snapshot(self) -> Iterable[SubmitRunInfo]:
+        return self._snapshot(self._try_fetch, self._crawl)
+
+    def _submit_runs_by_hour(self, start_date: datetime, end_date: datetime) -> Iterable[BaseRun]:
+        for start, end in _generate_hourly_ranges(start_date, end_date):
+            logger.info(f"Fetching submit runs from {start} to {end}")
+            yield from self._ws.jobs.list_runs(
+                expand_tasks=True,
+                completed_only=True,
+                run_type=ListRunsRunType.SUBMIT_RUN,
+                start_time_from=_dt_to_ms(start),
+                start_time_to=_dt_to_ms(end),
+            )
+        logger.info(f"Finished fetching submit runs from {start_date} to {end_date}")
+
+    def _crawl(self, num_days_history: int = 30) -> Iterable[SubmitRunInfo]:
+        submit_runs = list(self._submit_runs_by_hour(datetime.now() - timedelta(days=num_days_history), datetime.now()))
+        return self._assess_job_runs(submit_runs)
+
+    def _try_fetch(self) -> Iterable[SubmitRunInfo]:
+        for row in self._fetch(f"SELECT * FROM {self._schema}.{self._table}"):
+            yield SubmitRunInfo(*row)
+
+    def _retrieve_hash_values_from_sql_task(self, task: SqlTask) -> list[str]:
+        hash_values = []
+        if task.file is not None:
+            hash_values.append(task.file.path)
+        if task.alert is not None and task.alert.alert_id is not None:
+            hash_values.append(task.alert.alert_id)
+        if task.dashboard is not None and task.dashboard.dashboard_id is not None:
+            hash_values.append(task.dashboard.dashboard_id)
+        if task.query is not None and task.query.query_id is not None:
+            hash_values.append(task.query.query_id)
+        return [str(value) for value in hash_values if value is not None]
+
+    def _retrieve_hash_values_from_git_source(self, source: GitSource) -> list[str]:
+        hash_values = []
+        if source.git_url is not None:
+            hash_values.append(source.git_url)
+        return [str(value) for value in hash_values if value is not None]
+
+    def _retrieve_hash_values_from_task(self, task: RunTask) -> list[str]:
```

Comment:
This method is too long and too complex, break it down

Reply:
agree, splitted